packages/kmod-5.10-nvidia/nvidia-persistenced.service.in

```diff
@@ -0,0 +1,40 @@
+# NVIDIA Persistence Daemon Init Script -- Modified
+#
+# Copyright (c) 2013 NVIDIA Corporation
+#
+# Permission is hereby granted, free of charge, to any person obtaining a
+# copy of this software and associated documentation files (the "Software"),
+# to deal in the Software without restriction, including without limitation
+# the rights to use, copy, modify, merge, publish, distribute, sublicense,
+# and/or sell copies of the Software, and to permit persons to whom the
+# Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+# DEALINGS IN THE SOFTWARE.
+#
+# This is a sample systemd service file, designed to show how the NVIDIA
+# Persistence Daemon can be started.
+#
+
+[Unit]
+Description=NVIDIA Persistence Daemon
+Requires=load-kernel-modules.service
+After=load-kernel-modules.service
+
+[Service]
+Type=forking
+# Run the NVIDIA System Management Interface to create the device nodes.
+ExecStartPre=__NVIDIA_BINDIR__/nvidia-smi
```

Comment:
According to the `nvidia-persistenced` man pages:

> The daemon indirectly utilizes nvidia-modprobe via the nvidia-cfg library to load the NVIDIA kernel module and create the NVIDIA character device files after the daemon has dropped its root privileges, if it had any to begin with. If nvidia-modprobe is not installed, the daemon may not be able to start properly if it is not run with root privileges. 

So I assume you called `nvidia-smi` to circumvent the problem of creating the devices. Do you have the same problem if you run `nvidia-persistenced` as UID 0?

Reply:
When I run `nvidia-persistenced` with UID 0 it seems to create the device files just fine. It's only when I provide a user that it struggles. I'm guessing having the daemon run under a user is a better security practice, so that's why I've gone to some lengths to do the `--user` approach.

Reply:
`nvidia-smi` calls are a bit expensive. It won't only create the devices, but  it actually queries them so this operation may slowdown the boot time.

Regarding changing the user of the daemon, the man pages mention that `nvidia-modprobe` will be executed to create the devices but this binary has to be installed as `setuid` and at `/usr/bin/` :

> The application nvidia-modprobe ('/usr/bin/nvidia-modprobe') is installed as setuid root and is used to load the NVIDIA kernel module and create the '/dev/nvidia*' device nodes by processes that don't run with sufficient privileges to do those things themselves.

So even if we provide `nvidia-modprobe` (which is already provided in the run archive, and we could include it), the devices are still created by UID 0 (the privilege we were trying to drop).

I think that changing the UID of the daemon doesn't really get us too much, and it should be OK to just run the daemon as UID 0. Changing the UID will require us to include one more binary which doesn't seem necessary when running as UID 0, since the binary itself requires to be `setuid` to be able to create the devices. And this binary isn't even used today in containers, so we don't get too much value out of it by adding it to the distro.

If we are concerned about the privileges of the daemon, we could use systemd directives to restrict its privileges, e.g.:

```
NoNewPrivileges=true
```

https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#NoNewPrivileges=

You could even play around with:

https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#SystemCallFilter=

However, we don't know what system calls the daemon (or any of the libraries it depends on) run, it will be rather hard to nail down the correct filter.